README.md
![Foundry Core Compatible Version](https://img.shields.io/badge/dynamic/json.svg?url=https%3A%2F%2Fraw.githubusercontent.com%2Fspetzel2020%2Fspell-better%2Fmaster%2Fmodule.json&label=Foundry%20Version&query=$.compatibleCoreVersion&colorB=orange)
![Latest Release Download Count](https://img.shields.io/badge/dynamic/json?label=Downloads@latest&query=assets%5B1%5D.download_count&url=https%3A%2F%2Fapi.github.com%2Frepos%2Fspetzel2020%2Fspell-better%2Freleases%2Flatest)
![Forge Installs](https://img.shields.io/badge/dynamic/json?label=Forge%20Installs&query=package.installs&suffix=%25&url=https%3A%2F%2Fforge-vtt.com%2Fapi%2Fbazaar%2Fpackage%2Fspell-better&colorB=4aa94a)
[![ko-fi](https://img.shields.io/badge/-buy%20me%20a%20coffee-%23FF5E5B)](https://ko-fi.com/spetzel)

# Spetzel's Spell Better
The ultimate detachable spellbook for Wizards and other prepared-spell full casters (Druids and Clerics).

- Switch your character sheet to the Spell Better sheet for a full OGL-5e sheet with Spell Better spellbook
- OR Detach the Spell Better Spellbook by clicking the Spellbook tab from any supported character sheet
- See a one-line spell-slot summary as well as how many spells you can and have prepared
- Filter your spell list by level, school, usage, duration, prepared, concentration, and ritual
- Create custom Views like <strong>Utility Spells</strong>, <strong>AoE Nukes</strong>, <strong>Traveling</strong>, or <strong>Dungeoneering</strong>
- Create separate Spellbooks like <strong>Wanted</strong>, <strong>Found Scrolls</strong>, or <strong>Green Spellbook(deciphered)</strong>
- Easily print your Spellbook with custom views and sub-spellbooks, for quick reference at the screen or at the table

## Summary
* **Authors**: Discord: Spetzel#0103, [Calego#0914](https://github.com/ElfFriend-DnD)
* **Version**: 0.8.2
* **Foundry VTT Compatibility**: 0.7.2-0.7.9
* **System Compatibility (If applicable)**: dnd5e
* **Translation Support**: en

Spell Better is a forked version of https://github.com/ElfFriend-DnD/foundryvtt-5eOGLCharacterSheet, itself "heavily inspired by the Official 5e Character Sheet and the Roll20 default 5e Sheet. This sheet is chaotic and packed with information all on one screen, but it does have the advantage of having some muscle memory if you're coming from either pen and paper or Roll20."

## Installation
1. Go to the Add-on Modules tab in Foundry Setup
2. Click Install Module and search for **Spetzel's Spell Better** OR paste this link: `https://github.com/spetzel2020/spell-better/releases/download/latest/module.json`
3. Open your world and go to Settings>Manage Modules and enable Spetzel's Spell Better

## Gallery
[<img src="readme-img/spell-better-filters.png" width="30%"></img>](readme-img/spell-better-filters.png)

[<img src="readme-img/spell-better-custom-spellbook.png" width="30%"></img>](readme-img/spell-better-custom-spellbook.png)

See the github for the [5e-OGL-Character-Sheet](https://github.com/ElfFriend-DnD/foundryvtt-5eOGLCharacterSheet) for gallery of non-spellbook portions of the sheet.

Click to view bigger.

## Key Features & Changes

### Actions Area
Dead center of the screen this is the place where all of the "combat-important" (damage-dealing) spells and items live. Option in settings to limit spells to only Cantrips.

### Foundry-style Spellbook
It's not as familiar for a Roll20 user, but believe me when I say it's improved in almost every way.

## Options

| **Name**                           | Description                                                                                             |
| ---------------------------------- | ------------------------------------------------------------------------------------------------------- |
| **Limit Actions to Cantrips**      | Instead of showing all spells that deal damage in the Actions panel, limit it to only cantrips.         |
| **Add Icons to Inventory**         | Adds icons to all items in the inventory section, might make itemes with limited charges display oddly. |
| **Hide Category with no spells**   | If a category (for example, "7th Level") has no spells in it, hide it                                   |
| **Substitute for Spellbook**       | Open a pop-up version of the Spell Better Spellbook by clicking the Spellbook tab on other sheets       |

This sheet respects the 5e System setting: "Disable Experience Tracking"

### Compatibility

See Compatibility for https://github.com/ElfFriend-DnD/foundryvtt-5eOGLCharacterSheet

## Acknowledgements

Forked from https://github.com/ElfFriend-DnD/foundryvtt-5eOGLCharacterSheet (refer there for acknowledgements)


CHANGELOG.md
# RELEASE NOTES

## 0.8 26-Jan-2021
- New "Filter" category: Filter your known spells using one of the same filters as in the Filter list
- Fixed: Custom categories were being stored/not-stored inconsistently
- Fixed: Spells in custom Views/Spellbooks sometimes disappeared
    - NOTE: Your spellbook will be migrated to the new 0.8 format which may result in "missing" spells reappearing
- Fixed: A spell could not be in multiple Views
- Fixed Bugs: Below
## 0.7.5 25-Jan-2021
- Fixed: Was showing the Filters dropdown for View and Spellbook
- Fixed: Weren't able to actually create a Filter category replicating the standard ones

## 0.7.4 23-Jan-2021
- Fixed #4: Multiple collapsible groups with saved filters, or draggable groups like Inventory+
- Fixed #11, #8: Added Innate spell category and adjust spell uses/slots correctly
- Fixed #13: If a category is hidden (because of the setting to not show categories with 0 spells) it will still get considered for the Move Up/Down flags
- Fixed #9:  Block favTab for the Plugin because this is only a Spell Sheet
## 0.7.3 9-Jan-2021
- Fixed: Show/hide a spell category was very slow
- Fixed: Categories weren't sorted when you first loaded the spellbook
- Fixed: Substitute for Spellbook setting is now per-client
## 0.7.2 6-Jan-2021
- Fixed: Failing completely to open Spell Sheet for new actors (because of missing MODULE_VERSION)
## 0.7.0 5-Jan-2021
- Renumbered to be ahead of ogl sheet
- Switch to View and Spellbook options for Custom
## 0.5.3 4-Jan-2021
- Option to pop-up SpellBetter from other sheet's Spellbook tab
## 0.5.2 1-Jan-2021
- Add standard categories
- Show/hide filter controls
## 0.5.1 21-Dec-2020
- Incorporated Inventory+ into Spell Better (converted to apply to spells)
- Create new spell in special category
## 0.5.0 
- Forked from ogl-5e-character-sheet???



modules/helpers.js
/*
16-Dec-2020     Added hasAttack and hasDamage (copies of systems/dnd53/module/item/sheet.js#ItemSheet5e.hasAttack)

*/

import { MODULE_ID } from "./constants.js";
export function log(force, ...args) {
    if (force || CONFIG[MODULE_ID].debug === true) {
        console.log(MODULE_ID, '|', ...args);
    }
}

export function hasAttack(spell) {
    return ["mwak", "rwak", "msak", "rsak"].includes(spell?.data?.actionType);
}
export function hasDamage(spell) {
    return (spell?.data?.damage?.parts.length > 0);
}


export function getActivationType(activationType) {
    switch (activationType) {
        case 'action':
        case 'bonus':
        case 'reaction':
            return activationType;
        default:
            return 'special';
    }
}
export function getWeaponRelevantAbility(itemData, actorData) {
    if (!('ability' in itemData)) {
        return null;
    }
    const { ability, weaponType, properties } = itemData;
    // Case 1 - defined directly by the itemData
    if (ability) {
        return ability;
    }
    // Case 2 - inferred from actorData
    if (actorData) {
        // Melee weapons - Str or Dex if Finesse (PHB pg. 147)
        if (['simpleM', 'martialM'].includes(weaponType)) {
            if (properties.fin === true) {
                // Finesse weapons
                return actorData.abilities['dex'].mod >= actorData.abilities['str'].mod ? 'dex' : 'str';
            }
            return 'str';
        }
        // Ranged weapons - Dex (PH p.194)
        if (['simpleR', 'martialR'].includes(weaponType)) {
            return 'dex';
        }
        return 'str';
    }
    // Default null
    return null;
}


modules/Category.js
/*
29-Dec-2020     0.5.1t: Only save the customCategories and merge with the standardCategories (then we don't have to upgrade standard categories)              
                Split out from Inventory+ForSpells.js for creating and deleting categories
1-Jan-2021      0.5.1aa: getData(): includeInStandard -> showOnlyInCategory      
2-Jan-2021      0.5.1ac: Handle editCategory       
                0.5.1ad: Create choices structure to make mapping suitable for selectOptions Handlebars
                0.5.1ae findFilterSet(): Quick hack to check the constructed choice field too    
3-Jan-2021      0.5.2a: Switch to SPELL_BETTER.labelFilterSets format more consistent with selectOPtions
5-Jan-2021      0.5.3e: Change to a simple View or Spellbook selector
                0.5.3k: Sort after adding a new category and before saving
10-Jan-2021     0.7.3: _updateObject(): Don't save after sorting (done in sortCategories())       
25-Jan-2021     0.7.5: activateListeners(): Add on select.change() so that we enable/disable Filters depending on whether it's a Filter type         
26-Jan-2021     0.8.0: Switch to referencing spells in category rather than category in spells
                Filter: No effect (automatic)
                Spellbook/View: For now don't change the list of spells and see what happens
27-Jan-2021     0.8.1: Copy spellIds from old to new category (even if you change just the name)                
*/

import { MODULE_ID, SPELL_BETTER } from './constants.js';

export class CategorySheet extends FormApplication {
    constructor(categoryKey, options, inventoryPlusForSpells) {
        const categoryData = categoryKey ? inventoryPlusForSpells?.allCategories[categoryKey] : null;
        super(categoryData, options);
        this.categoryKey = categoryKey;
        this.inventoryPlusForSpells = inventoryPlusForSpells;

        //0.5.1ad: Convert filter choices to mapping suitable for selectOPtions Handlebars helper
        this.choices = {};
        for (const [filterSetKey, labelFilters] of Object.entries(SPELL_BETTER.labelFilterSets)) {
            this.choices[filterSetKey] = {};
            for (const [labelFilterKey, labelFilter] of Object.entries(labelFilters)) {
                this.choices[filterSetKey][labelFilterKey] = labelFilter.name;
            }
        }

    }

    static get defaultOptions() {
        return mergeObject(super.defaultOptions, {
            id: "create-category",
            title:  game.i18n.localize("SPELL_BETTER.NewCategory.TITLE"),
            classes: ["sheet","combat-sheet"],
            template: 'modules/spell-better/templates/categoryDialog.hbs',
            width: 420
        });
    }

    async getData(options) {
        //0.5.1ae: Populate selectedArray to pre-select for existing categories

        let selectedFilters = [];
        if (this.object) {
            for (const [filterSetKey, filters] of Object.entries(this.object?.labelFilterSets)) {
                const sbEntries = Object.entries(SPELL_BETTER.labelFilterSets[filterSetKey]);
                const labels = sbEntries.filter(entry => filters.includes(entry[1].label)).map(entry => entry[0]);
                selectedFilters = selectedFilters.concat(labels);
            }
        }
        const types = {
            view:  "SPELL_BETTER.NewCategory.Type.VIEW",
            filter: "SPELL_BETTER.NewCategory.Type.FILTER",            
            spellbook: "SPELL_BETTER.NewCategory.Type.SPELLBOOK"
        }
        const hideFilters = (this.object?.categoryType !== "filter");
        const templateData = {
            categoryKey : this.categoryKey,
            isCustom: this.object?.isCustom,
            label: this.object?.label,
            types: types,
            selectedType: this.object?.categoryType,
            hideFilters: hideFilters,      //trigger this from the type
            labelFilterSets : this.choices ,
            selectedFilters: selectedFilters
        }
        return templateData;
    }

    activateListeners(html) {
        super.activateListeners(html);

        //Disable the filters unless you pick "Filters"
        html.find("select#categoryType").change(ev => {
            const categoryType = ev.currentTarget.value;
            html.find("select#filter").prop("disabled",(categoryType !== "filter"));
        });
    }

    async _updateObject(event, formData) {

        //If this is a new category, create it
        let categoryKey = formData.categoryKey;
        if (!categoryKey) {
            //Validation
            if (!formData.label || formData.label === '') {
                ui.notifications.error('Could not create Category as no name was specified');
                return;
            }
            //Creating a new key should be encapsulated here
            categoryKey = this.inventoryPlusForSpells?.generateCategoryId();
        } 

        let newCategoryData = {
            label: formData.label,
            labelFilterSets: {},
            type: "spell",
            isCustom: true,
            canCreate: false,   //Not the place to create new spells
            canPrepare: true,
            isCollapsed: false,
            categoryType: "view",
            order: this.inventoryPlusForSpells?.getHighestSortFlag()-1  //Otherwise we would have to sort and decorate the move up/down
        }

        for (let [input, value] of Object.entries(formData)) {
            if (value === null || value === undefined || value === "") {
                delete newCategoryData[input];
            } else {
                //If this is (one or more) filters, then recover the filterSets (because we use those to correctly apply)
                if (input === "filter") {
                    const {filterSetKey,label} = CategorySheet.findFilterSet(value);
                    if (filterSetKey) {
                        if (!newCategoryData.labelFilterSets[filterSetKey]) newCategoryData.labelFilterSets[filterSetKey] = [];
                        newCategoryData.labelFilterSets[filterSetKey].push(label);
                    }
                } else {
                    newCategoryData[input] = value;
                }
            }
        }

        //0.5.3e DEPRECATED: Set appropriate templateFlags (for new or dragged spells)
        //This is NOT an array - spells can have multiple View categories, but Views and SPellbooks just have/apply one
        //0.8.0 DEPRECATED: No longer are setting flags on spells so don't need templateFlags
        //newCategory.templateFlags = {category: categoryKey};
        //Don't have set filterFlags because we use the categoryType setting
        if (this.inventoryPlusForSpells) {
            //0.8.0 If type is now "filter" there should be no saved spells (shouldn't be anyway since we're building from scratch)
            if (newCategoryData.categoryType === "filter") {
                delete newCategoryData.spellIds;
            } else {
                //Make sure we don't silently keep old filters
                newCategoryData.labelFilterSets = {};
                newCategoryData.spellIds = this.inventoryPlusForSpells.allCategories[categoryKey]?.spellIds;
                //If you switched from View->Spellbook, then those spells will be removed from other Views
                //If you switch from Spellbook -> View, then those spells are now part of your Known spells
            }
            this.inventoryPlusForSpells.allCategories[categoryKey] = newCategoryData;
            //0.5.3k: Sort after adding a new category and before saving
            //0.7.3: Also saves
            this.inventoryPlusForSpells.sortCategories();
        }
    }

    static findFilterSet(filter) {
        if (!filter) return null;
        for (const [filterSetKey, filters] of Object.entries(SPELL_BETTER.labelFilterSets)) {
            //0.5.1ae Quick hack to check the constructed choice field too
            for (const [fKey, f] of Object.entries(filters)) {
                if (fKey === filter) return {filterSetKey, label: f.label};
            }
        }
        return null;
    }

    
}//end class CategorySheet


modules/settings.js
/*
4-Jan-2021  0.5.3: Add substituteForSpellbook (whether we should pop up Spell Better instead of a regular spellbook tab)
9-Jan-2021 0.7.3a: Substitute for spellbook is a client setting so individual players can choose it

*/

import { SPELL_BETTER, MODULE_ID } from './constants.js';

export const registerSettings = function () {
  CONFIG[MODULE_ID] = { debug: false };

  // Register any custom module settings here
  game.settings.register(MODULE_ID, SPELL_BETTER.limitActionsToCantrips, {
    name: 'SPELL_BETTER.limitActionsToCantrips',
    default: false,
    type: Boolean,
    scope: 'client',
    config: true,
    hint: 'SPELL_BETTER.limitActionsToCantripsHint',
  });

  game.settings.register(MODULE_ID, SPELL_BETTER.showIconsOnInventoryList, {
    name: 'SPELL_BETTER.showIconsOnInventoryList',
    default: false,
    type: Boolean,
    scope: 'client',
    config: true,
    hint: 'SPELL_BETTER.showIconsOnInventoryListHint',
  });

  game.settings.register(MODULE_ID, SPELL_BETTER.expandedLimited, {
    name: 'SPELL_BETTER.expandedLimited',
    default: false,
    type: Boolean,
    scope: 'world',
    config: true,
    hint: 'SPELL_BETTER.expandedLimitedHint',
  });
    game.settings.register(MODULE_ID, SPELL_BETTER.hideCategoryWithNoSpells, {
        name:  "SPELL_BETTER.HideCategoryWithNoSpells.NAME",
        default: true,
        type: Boolean,
        scope: 'world',
        config: true,
        hint:  "SPELL_BETTER.HideCategoryWithNoSpells.HINT",
     });
    game.settings.register(MODULE_ID, SPELL_BETTER.substituteForSpellbook, {
        name:  "SPELL_BETTER.SubstituteForSpellbook.NAME",
        default: true,
        type: Boolean,
        scope: 'client',
        config: true,
        hint:  "SPELL_BETTER.SubstituteForSpellbook.HINT",
     });
};


modules/preloadTemplates.js
import { MODULE_ID } from './constants.js';

export const preloadTemplates = async function () {
  const templatePaths = [
    `modules/${MODULE_ID}/assets/armor-class.hbs`,
    `modules/${MODULE_ID}/assets/strength.hbs`,
    `modules/${MODULE_ID}/templates/character-sheet-ltd.hbs`,
    `modules/${MODULE_ID}/templates/character-sheet.hbs`,
    `modules/${MODULE_ID}/templates/parts/actor-actions-list.hbs`,
    `modules/${MODULE_ID}/templates/parts/actor-health-armor.hbs`,
    `modules/${MODULE_ID}/templates/parts/actor-skills.hbs`,
    `modules/${MODULE_ID}/templates/parts/actor-features.hbs`,
    `modules/${MODULE_ID}/templates/parts/actor-inventory.hbs`,
    `modules/${MODULE_ID}/templates/parts/actor-spellbook.hbs`,
    `modules/${MODULE_ID}/templates/parts/actor-traits.hbs`,
    `modules/${MODULE_ID}/templates/parts/sheet-header.hbs`,
    `modules/${MODULE_ID}/templates/parts/spellbook-sheet-header.hbs`,
    `modules/${MODULE_ID}/templates/spellbook-sheet-holder.hbs`,
  ];

  return loadTemplates(templatePaths);
};


modules/constants.js
/*
29-Dec-2020 0.5.1:  Add isCustom flag to added categories
30-Dec-2020 0.5.1x: OPtion to not display standard categories with no spells
2-Jan-2021  0.5.1ad: Add choice to labelFilterSets to make using selectOptions easier - they are converted into property names
                        (if not present then use filter value instead)
3-Jan-2021  0.5.2a: Switch labelFilterSets and standard categories to a format more consistent with foundry.js#selectOptions
            0.5.2b: SPELL_BETTER.standardCategories: Separate labelFilters and categoryFilters; add showOnlyNoCategory
4-Jan-2021  0.5.3: Substitute for the regular spellbook tab on supported sheets      
5-Jan-2021  0.5.3e: Use viewOrSpellbook choice    
7-Jan-2021  0.7.2: Updated MODULE_VERSION  
9-Jan-2021  0.7.3e: Remove Wanted 4th for release version
23-Jan-2021 0.7.4a: Add Innate category 
27-Jan-2021: 0.8.1: Make Ritual fully standard (can't change its name); remove template_flags
28-Jan-2021 0.8.2: Add at-will standard category (like innate)

*/


export const MODULE_ID = 'spell-better';
export const MODULE_VERSION = "0.8.3";

//SPell Better 0.5.0: The filter lists have to be spelled out to match what is in the internal labels
export const SPELL_BETTER = {
    limitActionsToCantrips : 'limit-actions-to-cantrips',
    showIconsOnInventoryList : 'show-icons-on-inventory-list',
    expandedLimited : 'expanded-limited',
    hideCategoryWithNoSpells : "hide-category-with-no-spells",
    substituteForSpellbook: "substitute-for-spellbook",
    labelFilterSets :  {
        castingTimes : {
            _1_Action : {label: "1 Action", name: "DND5E.Action"},         
            _1_Bonus_Action : {label: "1 Bonus Action", name: "DND5E.BonusAction"},      
            _1_Reaction: {label: "1 Reaction", name: "DND5E.Reaction"},                      
            _1_M: {label: "1 M", name: "SPELL_BETTER._1Minute"},
            _10_M: {label: "10 M",name: "SPELL_BETTER._10Minutes"},
            _1_H: {label: "1 H", name: "SPELL_BETTER._1Hour"}
        },
        otherConcentration : {concentration: {label: "concentration",name: "DND5E.Concentration"}},
        otherRitual : {ritual: {label: "ritual",name: "DND5E.Ritual"}},
        otherPrepared : {
            prepared: {label: "prepared",name: "DND5E.Prepared"},
            innate: {label: "innate", name: "SPELL_BETTER.Innate"}
        },
        durations : {
            instantaneous: {label: "Instantaneous", name: "DND5E.TimeInst"},                         
            _1_Round: {label: "1 Rounds",name: "SPELL_BETTER._1Round"},
            _1_Minute: {label: "1 Minutes", name: "SPELL_BETTER._1Minute"},
            _10_Minutes: {label: "10 Minutes",name: "SPELL_BETTER._10Minutes"},
            _1_Hours: {label: "1 Hours", name: "SPELL_BETTER._1Hour"}
        },
        schools : {
            abjuration: {label: "Abjuration", name: "DND5E.SchoolAbj"},
            conjuration: {label: "Conjuration", name: "DND5E.SchoolCon"},
            divination: {label: "Divination", name: "DND5E.SchoolDiv"},
            enchantment: {label: "Enchantment", name: "DND5E.SchoolEnc"},
            evocation: {label: "Evocation", name: "DND5E.SchoolEvo"},
            illusion: {label: "Illusion", name: "DND5E.SchoolIll"},
            necromancy: {label: "Necromancy", name: "DND5E.SchoolNec"},
            transmutation: {label: "Transmutation", name: "DND5E.SchoolTrs"}
        },
        levels : {
            cantrip: {label: "Cantrip", name: "DND5E.SpellLevel0"},
            "_1st_Level": {label: "1st Level", name: "SPELL_BETTER.SpellLevel1"},
            "_2nd_Level": {label: "2nd Level", name: "SPELL_BETTER.SpellLevel2"},
            "_3rd_Level": {label: "3rd Level", name: "SPELL_BETTER.SpellLevel3"},
            "_4th_Level": {label: "4th Level", name: "SPELL_BETTER.SpellLevel4"},
            "_5th_Level": {label: "5th Level", name: "SPELL_BETTER.SpellLevel5"},
            "_6th_Level": {label: "6th Level", name: "SPELL_BETTER.SpellLevel6"},
            "_7th_Level": {label: "7th Level", name: "SPELL_BETTER.SpellLevel7"},
            "_8th_Level": {label: "8th Level", name: "SPELL_BETTER.SpellLevel8"},
            "_9th_Level": {label: "9th Level",  name: "SPELL_BETTER.SpellLevel9"}
        }
    },
    categoriesVersion_key : "categoriesVersion",
    categoriesVersion : MODULE_VERSION, 
    categories_key : "categories",
    standardCategories : { 
        //.prop is needed for now for display     
        atwill: { label: "DND5E.SpellPrepAtWill", categoryType: "filter", 
                templateItemData: {level: 0},
                labelFilterSets: {
                    otherPrepared: ["atwill"]
                }, 
                /* Will be overridden by actual spell level stats */
                prop: "atwill", canCreate: true, canPrepare: false, slots: "&infin;", uses: "&infin;", usesSlots: false,
                order: -20, type: "spell",  isCollapsed: true },    
        innate: { label: "SPELL_BETTER.Innate", categoryType: "filter", 
                templateItemData: {level: 0},
                labelFilterSets: {
                    otherPrepared: ["innate"]
                }, 
                /* Will be overridden by actual spell level stats */
                prop: "innate", canCreate: true, canPrepare: false, slots: "&infin;", uses: "&infin;", usesSlots: false,
                order: -10, type: "spell",  isCollapsed: true },
        spell0: { label: "DND5E.SpellLevel0", categoryType: "filter", 
                templateItemData: {level: 0},
                labelFilterSets: {
                    levels: ["Cantrip"]
                }, 
                /* Will be overridden by actual spell level stats */
                prop: "spell0", canCreate: true, canPrepare: false, slots: "&infin;", uses: "&infin;", usesSlots: false,
                order: 0, type: "spell",  isCollapsed: true },
        spell1: { label: "SPELL_BETTER.SpellLevel1", categoryType: "filter", 
                templateItemData: {level: 1},
                labelFilterSets: {
                    levels: ["1st Level"]
                },         
                prop: "spell1", canCreate: true, canPrepare: true, slots: "4", uses: "4", usesSlots: true,    
                order: 10, type: "spell",  isCollapsed: false },     
        spell2: { label: "SPELL_BETTER.SpellLevel2", categoryType: "filter", 
                templateItemData: {level: 2},
                labelFilterSets: {
                    levels: ["2nd Level"]
                },             
                prop: "spell2", canCreate: true, canPrepare: true, slots: "3", uses: "3", usesSlots: true,       
                order: 20, type: "spell",  isCollapsed: false },
        spell3: { label: "SPELL_BETTER.SpellLevel3", categoryType: "filter", 
                templateItemData: {level: 3},
                labelFilterSets: {
                    levels: ["3rd Level"],
                },   
                prop: "spell3", canCreate: true, canPrepare: true, slots: "3", uses: "3", usesSlots: true,                
                order: 30,  type: "spell",  isCollapsed: false },
        spell4: { label: "SPELL_BETTER.SpellLevel4", categoryType: "filter", 
                templateItemData: {level: 4},
                labelFilterSets: {
                    levels: ["4th Level"]
                },       
                prop: "spell4", canCreate: true, canPrepare: true, slots: "3", uses: "3", usesSlots: true,                
                order: 40,  type: "spell",  isCollapsed: false },          
        spell5: { label: "SPELL_BETTER.SpellLevel5", categoryType: "filter", 
                templateItemData: {level: 5},
                labelFilterSets: {
                    levels: ["5th Level"]
                },       
                prop: "spell5", canCreate: true, canPrepare: true, slots: "3", uses: "3", usesSlots: true,                
                order: 50,  type: "spell",  isCollapsed: false },
        spell6: { label: "SPELL_BETTER.SpellLevel6",categoryType: "filter", 
                templateItemData: {level: 6},
                labelFilterSets: {
                    levels: ["6th Level"]
                },                 
                prop: "spell6", canCreate: true, canPrepare: true, slots: "1", uses: "1", usesSlots: true,                
                order: 60,  type: "spell",  isCollapsed: false },
        spell7: { label: "SPELL_BETTER.SpellLevel7", categoryType: "filter", 
                templateItemData: {level: 7},
                labelFilterSets: {
                    levels: ["7th Level"]
                },              
                prop: "spell7", canCreate: true, canPrepare: true, slots: "1", uses: "1", usesSlots: true,                
                order: 70,  type: "spell",  isCollapsed: false },
        spell8: { label: "SPELL_BETTER.SpellLevel8", categoryType: "filter", 
                templateItemData: {level: 8},
                labelFilterSets: {
                    levels: ["8th Level"]
                },                 
                prop: "spell8", canCreate: true, canPrepare: true, slots: "1", uses: "1", usesSlots: true,                
                order: 80,  type: "spell",  isCollapsed: false },
        spell9: { label: "SPELL_BETTER.SpellLevel9", categoryType: "filter", 
                templateItemData: {level: 9},
                labelFilterSets: {
                    levels: ["9th Level"]
                },                
                prop: "spell9", canCreate: true, canPrepare: true, slots: "1", uses: "1", usesSlots: true,                
                order: 90,  type: "spell",  isCollapsed: false },
        rituals: { label: "SPELL_BETTER.Rituals", categoryType: "filter", 
                templateItemData: {level: 1},
                labelFilterSets: {
                    otherRitual: ["ritual"]
                },         
                canCreate: false, canPrepare : false,           
                order: 900, usesSlots: false, type: "spell",  isCollapsed: true },                                                                                                                                                      
        wanted: { label: "SPELL_BETTER.Category.Wanted", isCustom: true, categoryType: "spellbook", 
                templateItemData: {level: 9}, 
                labelFilterSets: {}, 
                canCreate: true, canPrepare : false,           
                order: 1000, usesSlots: false, type: "spell",  isCollapsed: true },
        //ALL category to make sure all spells are accessible              
        all:    {label: "SPELL_BETTER.Category.All", categoryType: "all", 
                templateItemData: {level: 9},
                labelFilterSets: {},    
                canCreate: true, canPrepare : true,                     
                order: 9000, usesSlots: false, type: "spell",  isCollapsed: true }
    }
}
